### DISCUSSION TITLE
Package not found

### DISCUSSION THREAD

**Discussion by VishalPatel43** [LearnLib/learnlib#101] [General]
After setup the stable version of the learning there are new examples which have import package not found in the library
examples package (how to integrate this package used in the given example)

Library in examples **--> src --> main --> java --> de --> learnlib --> examples --> bbc** 

**Example1.java**

import de.learnlib.examples.LearningExample.DFALearningExample;
import de.learnlib.examples.dfa.ExampleTinyDFA;

**Example2.java**

import de.learnlib.examples.LearningExample.MealyLearningExample;
import de.learnlib.examples.mealy.ExampleTinyMealy;

**Examle3.java**

import de.learnlib.examples.LearningExample.MealyLearningExample;
import de.learnlib.examples.mealy.ExampleTinyMealy;

**Example4.java**

import de.learnlib.examples.LearningExample.DFALearningExample;
import de.learnlib.examples.dfa.ExampleTinyDFA;

Library in examples **--> src --> main --> java --> de --> learnlib --> examples**

**Example3.java**

import de.learnlib.examples.Example2.BoundedStringQueue;

**Comment by mtf90** [comment]
Try adding
```xml
<dependency>
    <groupId>de.learnlib.testsupport</groupId>
    <artifactId>learnlib-learning-examples</artifactId>
    <version>0.16.0</version>
</dependency>
```
Since these examples do not contain functional library code, they are not part of the distribution artifact.

**Comment by VishalPatel43** [comment]
I was trying to run this example and its produce this error

`/* Copyright (C) 2013-2020 TU Dortmund
 * This file is part of LearnLib, http://www.learnlib.de/.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.main.learnlibexample.bbc;

import java.util.function.Function;

import de.learnlib.acex.analyzers.AcexAnalyzers;
import de.learnlib.algorithms.ttt.dfa.TTTLearnerDFA;
import de.learnlib.api.algorithm.LearningAlgorithm.DFALearner;
import de.learnlib.api.logging.LoggingPropertyOracle;
import de.learnlib.api.oracle.EquivalenceOracle.DFAEquivalenceOracle;
import de.learnlib.api.oracle.InclusionOracle.DFAInclusionOracle;
import de.learnlib.api.oracle.LassoEmptinessOracle;
import de.learnlib.api.oracle.MembershipOracle.DFAMembershipOracle;
import de.learnlib.api.oracle.OmegaMembershipOracle.DFAOmegaMembershipOracle;
import de.learnlib.api.oracle.PropertyOracle;

import de.learnlib.examples.LearningExample.DFALearningExample;
import de.learnlib.examples.dfa.ExampleTinyDFA;

import de.learnlib.oracle.emptiness.DFALassoEmptinessOracleImpl;
import de.learnlib.oracle.equivalence.DFABFInclusionOracle;
import de.learnlib.oracle.equivalence.DFACExFirstOracle;
import de.learnlib.oracle.equivalence.DFAEQOracleChain;
import de.learnlib.oracle.equivalence.DFAWpMethodEQOracle;
import de.learnlib.oracle.membership.SimulatorOmegaOracle.DFASimulatorOmegaOracle;
import de.learnlib.oracle.property.DFALassoPropertyOracle;
import de.learnlib.util.Experiment;
import net.automatalib.automata.fsa.DFA;
import net.automatalib.modelcheckers.ltsmin.ltl.LTSminLTLDFABuilder;
import net.automatalib.modelchecking.ModelCheckerLasso.DFAModelCheckerLasso;
import net.automatalib.util.automata.equivalence.DeterministicEquivalenceTest;
import net.automatalib.words.Alphabet;

/**
 * Runs a black-box checking experiment for a DFA.
 *
 * @author Jeroen Meijer
 */
public final class Example1 {

    /**
     * A function that transforms edges in an FSM source to actual input for a DFA.
     */
    public static final Function<String, Character> EDGE_PARSER = s -> s.charAt(0);

    private Example1() {}

    public static void main(String[] args) {

        DFALearningExample<Character> le = ExampleTinyDFA.createExample();

        // define the alphabet
        Alphabet<Character> sigma = le.getAlphabet();

        // create the DFA to be verified/learned
        DFA<?, Character> dfa = le.getReferenceAutomaton();

        // create an omega membership oracle
        DFAOmegaMembershipOracle<?, Character> omqOracle = new DFASimulatorOmegaOracle<>(dfa);

        // create a regular membership oracle
        DFAMembershipOracle<Character> mqOracle = omqOracle.getMembershipOracle();

        // create a learner
        DFALearner<Character> learner = new TTTLearnerDFA<>(sigma, mqOracle, AcexAnalyzers.LINEAR_FWD);

        // create a model checker
        DFAModelCheckerLasso<Character, String> modelChecker =
                new LTSminLTLDFABuilder<Character>().withString2Input(EDGE_PARSER).create();

        // create an emptiness oracle, that is used to disprove properties
        LassoEmptinessOracle.DFALassoEmptinessOracle<Character>
                emptinessOracle = new DFALassoEmptinessOracleImpl<>(omqOracle);

        // create an inclusion oracle, that is used to find counterexamples to hypotheses
        DFAInclusionOracle<Character> inclusionOracle = new DFABFInclusionOracle<>(mqOracle, 1.0);

        // create an LTL property oracle, that also logs stuff
        PropertyOracle.DFAPropertyOracle<Character, String> ltl = new LoggingPropertyOracle.DFALoggingPropertyOracle<>(
            new DFALassoPropertyOracle<>("letter==\"b\"", inclusionOracle, emptinessOracle, modelChecker));

        // create an equivalence oracle, that first searches for a counter example using the ltl properties, and next
        // with the W-method.
        @SuppressWarnings("unchecked")
        DFAEquivalenceOracle<Character> eqOracle = new DFAEQOracleChain<>(
                new DFACExFirstOracle<>(ltl),
                new DFAWpMethodEQOracle<>(mqOracle, 3));

        // create an experiment
        Experiment.DFAExperiment<Character> experiment = new Experiment.DFAExperiment<>(learner, eqOracle, sigma);

        // run the experiment
        experiment.run();

        // get the result
        final DFA<?, Character> result = experiment.getFinalHypothesis();

        // assert we have the correct result
        assert DeterministicEquivalenceTest.findSeparatingWord(dfa, result, sigma) == null;
    }
}
`

![Uploading Screenshot 2023-11-03 at 22.56.49.png…]()

**Reply by mtf90** [reply]
Unfortunatey, I cannot open your supposed screenshot, but I would assume the error is because you haven't installied LTSmin which is required by the Black-Box-Checking code/examples. Since black-box checking is not really relevant to your regexp/L*/NL* comparison, I wouldn't waste too much time for trying to get the example running.

**Reply by VishalPatel43** [reply]
Okay sir,  I understood 

**While I try try run below example which provide in automatalib --> 'SimpleBricsExample.java'

/* Copyright (C) 2013-2020 TU Dortmund
 * This file is part of AutomataLib, http://www.automatalib.net/.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.main.automatalibexample.brics;

import java.util.Arrays;
import java.util.List;

import dk.brics.automaton.Automaton;
import dk.brics.automaton.RegExp;
import net.automatalib.brics.AbstractBricsAutomaton;
import net.automatalib.brics.BricsNFA;
import net.automatalib.visualization.Visualization;
import net.automatalib.words.Word;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class SimpleBricsExample {

    private static final Logger LOGGER = LoggerFactory.getLogger(SimpleBricsExample.class);

    private SimpleBricsExample() {}

    public static void main(String[] args) {
        // Create a BRICS automaton from a regular expression ...
        RegExp r = new RegExp("ab+(c|d)*e?");
        Automaton a = r.toAutomaton();
        // ... and wrap it into the AutomataLib interfaces
        AbstractBricsAutomaton ba = new BricsNFA(a);

        // Then, display a DOT representation of this automaton
        Visualization.visualize(ba, true);

        // Test whether the following words are accepted
        List<Word<Character>> testWords = Arrays.asList(Word.fromString("abd"),
                                                        Word.fromString("abbc"),
                                                        Word.fromString("abbbbbde"),
                                                        Word.fromString("ade"));

        for (Word<Character> tw : testWords) {
            LOGGER.info("Output for {} is {}", tw, ba.computeOutput(tw));
        }
    }

}



<img width="1440" alt="Screenshot 2023-11-04 at 13 20 38" src="https://github.com/LearnLib/learnlib/assets/123478469/cebffbb0-e9d4-4daa-b676-886a822e0104">

**Comment by VishalPatel43** [comment]
I add this dependecy and error was gone but there is warning and where can I find the LOGGER output

```
 <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>1.7.5</version>
        </dependency>

        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-log4j12</artifactId>
            <version>1.7.5</version>
        </dependency>
```


<img width="1440" alt="Screenshot 2023-11-04 at 15 00 34" src="https://github.com/LearnLib/learnlib/assets/123478469/c1a8eeec-0add-4b22-b3aa-22b923245002">

**Reply by VishalPatel43** [reply]
I fix this with create the file log4j.xml (src/main/resources) for maven

first add the dependecy

```
        <dependency>
            <groupId>log4j</groupId>
            <artifactId>log4j</artifactId>
            <version>1.2.17</version> <!-- Use the latest version available -->
        </dependency>
```

log4j.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE log4j:configuration SYSTEM "log4j.dtd">
<log4j:configuration xmlns:log4j="http://jakarta.apache.org/log4j/">

    <appender name="console" class="org.apache.log4j.ConsoleAppender">
        <layout class="org.apache.log4j.PatternLayout">
            <param name="ConversionPattern" value="%d{yyyy-MM-dd HH:mm:ss} %-5p %c{1}:%L - %m%n" />
        </layout>
    </appender>

    <root>
        <priority value="debug" />
        <appender-ref ref="console" />
    </root>

</log4j:configuration>

```

**Reply by mtf90** [reply]
Great that you could solve the issue for yourself. LearnLib/AutomataLib uses the logging facade SLF4J which allows you to select a custom implementation at runtime. Depending on which you choose (e.g., log4j or logback), you may need a specific configuration file to configure it properly.

**Reply by mtf90** [reply]
Is this discussion resolved then?
